src/fs/data_compaction.cc

```diff
@@ -0,0 +1,397 @@
+        }
+    }
+
+    // It is quite common to call compaction with empty clusters to immediately free those clusters
+    auto compact_future = data_vecs.empty() ? now() : read_data_vectors(std::move(data_vecs));
+
+    return compact_future.finally([this] {
+        auto metadata_log_flush = _was_metadata_log_modified ? _metadata_log.flush_log() : now();
+        return metadata_log_flush.then([this] {
+            return parallel_for_each(_compacted_cluster_ids, [this](cluster_id_t cluster_id) {
+                auto cluster_it = _metadata_log._writable_data_clusters.find(cluster_id);
+                assert(cluster_it != _metadata_log._writable_data_clusters.end());
+                if (cluster_it->second.is_empty()) {
+                    return cluster_it->second.wait_for_all_readers_to_unlock().then(
+                            [this, cluster_id, cluster_it = std::move(cluster_it)] {
+                        // TODO: throw if read_only or allow it here?
+                        mlogger.debug("releasing cluster: {}", cluster_id);
+                        _metadata_log.free_writable_data_cluster(cluster_id);
+                    });
+                } else {
+                    mlogger.warn("cluster {} not empty after compaction - cannot free", cluster_id);
+                    _metadata_log.finish_writing_data_cluster(cluster_id);
+                    return now();
+                }
+            });
+        }).handle_exception([this](std::exception_ptr ptr) {
+            for (auto& cluster_id : _compacted_cluster_ids) {
+                _metadata_log.finish_writing_data_cluster(cluster_id);
+            }
+            mlogger.warn("Exception while flushing log after compaction: cannot free compacted clusters.");
+            return make_exception_future(std::move(ptr));
+        });
+    });
+}
+
+future<> data_compaction::read_data_vectors(std::vector<compacted_data_vec> data_vecs) {
+    // TODO: We could read the whole cluster when we read more data from it than some threshold (for example
+    //       more than 40%).
+    return do_with(device_reader(_metadata_log._device, _metadata_log._alignment), std::move(data_vecs),
+            [this](device_reader& reader, std::vector<compacted_data_vec>& data_vecs) {
+        return do_for_each(data_vecs, [&reader](compacted_data_vec& vec) {
+            return reader.read(vec._data.get_write(), vec._prev_disk_offset, vec._data.size()).then(
+                    [expected_read_len = vec._data.size()](size_t read_len) {
+                if (read_len != expected_read_len) {
+                    return make_exception_future<>(input_output_exception());
+                }
+                return now();
+            });
+        }).then([this, &data_vecs] {
+            return group_data_into_clusters(std::move(data_vecs));
+        });
+    });
+}
+
+future<> data_compaction::group_data_into_clusters(std::vector<compacted_data_vec> read_data_vecs) {
```

Comment:
This function is so long, that it would be nice to have it divided into smaller parts.